extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,222 @@
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+  }
+
+  private List<InputRow> parseMetricsData(final MetricsData metricsData)
+  {
+    return metricsData.getResourceMetricsList()
+            .stream()
+            .flatMap(resourceMetrics -> resourceMetrics.getInstrumentationLibraryMetricsList()
+                    .stream()
+                    .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
+                            .stream()
+                            .flatMap(metric -> parseMetric(metric, resourceMetrics.getResource().getAttributesList())
+                                    .stream())))
+            .collect(Collectors.toList());
+  }
+
+  private List<InputRow> parseMetric(Metric metric, List<KeyValue> resourceAttributes)
+  {
+    switch (metric.getDataCase()) {
+      case INT_SUM: {
+        return metric.getIntSum()
+                .getDataPointsList()
+                .stream()
+                .map(OpenTelemetryMetricsProtobufReader::intDataPointToNumberDataPoint)
+                .map(dataPoint -> parseNumberDataPoint(dataPoint, metric.getName(), resourceAttributes))
+                .collect(Collectors.toList());
+      }
+      case INT_GAUGE: {
+        return metric.getIntGauge()
+                .getDataPointsList()
+                .stream()
+                .map(OpenTelemetryMetricsProtobufReader::intDataPointToNumberDataPoint)
+                .map(dataPoint -> parseNumberDataPoint(dataPoint, metric.getName(), resourceAttributes))
+                .collect(Collectors.toList());
+      }
+      case SUM: {
+        return metric.getSum()
+                .getDataPointsList()
+                .stream()
+                .map(dataPoint -> parseNumberDataPoint(dataPoint, metric.getName(), resourceAttributes))
+                .collect(Collectors.toList());
+      }
+      case GAUGE: {
+        return metric.getGauge()
+                .getDataPointsList()
+                .stream()
+                .map(dataPoint -> parseNumberDataPoint(dataPoint, metric.getName(), resourceAttributes))
+                .collect(Collectors.toList());
+      }
+      // TODO Support HISTOGRAM and SUMMARY metrics
+      default:
+        throw new IllegalStateException("Unexpected value: " + metric.getDataCase());
+    }
+  }
+
+  private static NumberDataPoint intDataPointToNumberDataPoint(IntDataPoint dataPoint)
+  {
+    return NumberDataPoint.newBuilder()
+            .setTimeUnixNano(dataPoint.getTimeUnixNano())
+            .setAsInt(dataPoint.getValue())
+            .addAllAttributes(dataPoint.getLabelsList()
+                    .stream()
+                    .map(label -> KeyValue.newBuilder()
+                            .setKey(label.getKey())
+                            .setValue(AnyValue.newBuilder().setStringValue(label.getValue()))
+                            .build())
+                    .collect(Collectors.toList()))
+            .build();
+  }
```

Comment:
doing this conversion for every data point is going to be very expensive and create a lot of garbage, especially to copy all the attributes. Instead we should try to have specialized functions to handle each type

Reply:
looks like upgrading to 0.12 protos will get rid of this code and remove that concern.

Reply:
I'm not sure that we can do that yet, unless we can be sure that all the upstream collectors will be moved over to 0.11 by the time that this is deployed. Right now the satellites are still on v0.6. I suppose that we can prevent un-upgraded collectors from sending metrics to the shared topic and only point the upgraded collectors.